src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceRecordWriter.java

```diff
@@ -0,0 +1,119 @@
+ * Accepts <code>null</code> as a key, and CSVRecord as a value.
+ */
+public class SalesforceRecordWriter extends RecordWriter<NullWritable, CSVRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(SalesforceRecordWriter.class);
+
+  private BulkConnection bulkConnection;
+  private JobInfo jobInfo;
+  private ErrorHandling errorHandling;
+  private Long maxBytesPerBatch;
+  private Long maxRecordsPerBatch;
+  private List<BatchInfo> batchInfoList = new ArrayList<>();
+  private CSVBuffer csvBuffer;
+  private CSVBuffer csvBufferSizeCheck;
+
+  public SalesforceRecordWriter(TaskAttemptContext taskAttemptContext) throws IOException, AsyncApiException {
+    Configuration conf = taskAttemptContext.getConfiguration();
+    String jobId = conf.get(SalesforceSinkConstants.CONFIG_JOB_ID);
+
+    // these are already validated no need to validate again
+    errorHandling = ErrorHandling.fromValue(conf.get(SalesforceSinkConstants.CONFIG_ERROR_HANDLING)).get();
+    maxBytesPerBatch = Long.parseLong(conf.get(SalesforceSinkConstants.CONFIG_MAX_BYTES_PER_BATCH));
+    maxRecordsPerBatch = Long.parseLong(conf.get(SalesforceSinkConstants.CONFIG_MAX_RECORDS_PER_BATCH));
+
+    csvBuffer = new CSVBuffer(true);
+    csvBufferSizeCheck = new CSVBuffer(false);
+
+    AuthenticatorCredentials credentials = SalesforceConnectionUtil.getAuthenticatorCredentials(conf);
+    bulkConnection = new BulkConnection(Authenticator.createConnectorConfig(credentials));
+    jobInfo = bulkConnection.getJobStatus(jobId);
+  }
+
+  @Override
+  public void write(NullWritable key, CSVRecord csvRecord) throws IOException {
+    csvBufferSizeCheck.reset();
+    csvBufferSizeCheck.write(csvRecord);
+
+    if (csvBuffer.size() + csvBufferSizeCheck.size() > maxBytesPerBatch ||
+      csvBuffer.getRecordsCount() >= maxRecordsPerBatch) {
+      submitCurrentBatch();
+    }
+
+    csvBuffer.write(csvRecord);
+  }
+
+  private void submitCurrentBatch() throws IOException {
+    if (csvBuffer.getRecordsCount() != 0) {
+      InputStream csvInputStream = new ByteArrayInputStream(csvBuffer.getByteArray());
+      try {
+        BatchInfo batchInfo = bulkConnection.createBatchFromStream(jobInfo, csvInputStream);
+        batchInfoList.add(batchInfo);
+        LOG.info("Submitted a batch with batchId='{}'", batchInfo.getId());
+      } catch (AsyncApiException e) {
+        throw new RuntimeException("There was issue communicating with Salesforce", e);
+      }
+      csvBuffer.reset();
+    }
+  }
+
+  @Override
+  public void close(TaskAttemptContext taskAttemptContext) throws IOException {
+    submitCurrentBatch();
+
+    try {
+      SalesforceBulkUtil.awaitCompletion(bulkConnection, jobInfo, batchInfoList);
+      SalesforceBulkUtil.checkResults(bulkConnection, jobInfo, batchInfoList, errorHandling.equals(ErrorHandling.SKIP));
+    } catch (AsyncApiException e) {
+      throw new RuntimeException("There was issue communicating with Salesforce", e);
+    }
+
+    try {
```

Comment:
this should happen even if an AsyncApiException is caught above. Probably should have it in a finally block from above.

Reply:
fixed